earthfile2llb/converter.go

```diff
@@ -1765,6 +1815,50 @@ func (c *Converter) targetInputActiveOnly() dedup.TargetInput {
 	return c.mts.Final.TargetInput().WithFilterBuildArgs(activeBuildArgs)
 }
 
+// persistCache makes temporary cache directories permanent by writing their contents
+// from the cached directory to the persistent image layers at the same directory.
+// Used when the Target contains at least one `CACHE /my/directory` command.
+// Note that the RunOptions provided should contain at least all mounts corresponding to the cache direcories.
+func persistCache(srcState pllb.State, cacheDirs map[string]struct{}, opts []llb.RunOption, platform *specs.Platform) pllb.State {
+	// tmpDir is a backup directory where we can store the contents of the user's cache volume
+	// Is's name is sufficiently random that it should never collide with a user's work.
+	const tmpDir = "/earthly-tmp-a8cb9b0e-f285-4851-b00e-cd5b1ac6a499"
+	dest := srcState
+
+	// User may have multiple CACHE commands in a single target
+	for cacheDir := range cacheDirs {
+		// Copy the contents of the user's cache directory to the temporary backup.
+		// It's important to use DockerfileCopy here, since traditional llb.Copy()
+		// doesn't support adding mounts via RunOptions.
+		dest = llbutil.DockerfileCopy(
+			dest,
+			cacheDir,
+			tmpDir,
+			append(opts, llb.WithCustomName("persist cache directory")),
+			platform,
+		)
+
+		// Copy the contents from our tmp directory back to the
+		// same place as the user placed them in their cache directory
+		fa := pllb.Copy(
+			dest,
+			fmt.Sprintf("%s/*", tmpDir),
+			cacheDir,
+			&llb.CopyInfo{
+				FollowSymlinks:     true,
+				CreateDestPath:     true,
+				AllowWildcard:      true,
+				AllowEmptyWildcard: true,
+			},
+		)
+
+		// Remove the temporary backup after we're done copying from it.
+		dest = dest.File(fa.Rm(tmpDir))
```

Comment:
So this has 3*N layers:

1. Use copy+mounts to copy cache from mount into a tmp dir
2. Use an `llb.Copy` to copy from a tmp dir and into the previously mounted path
3. Rm tmp dir

I wonder if there's a way to reduce this as much as possible. For example, what if we could copy from the mount directly into the proper destination. Then we would need only a single copy and no rm.

I think we can achieve this by mounting the cache to a tmp location (as opposed to mounting in its usual location), and then copying from it to the real intended path.

Reply:
Interesting idea, I think in theory that should work. I'll give it a go

Reply:
Got this working, thanks for the suggestion!